cpp/src/arrow/type.cc

```diff
@@ -676,6 +676,16 @@ std::vector<std::string> Schema::field_names() const {
   return names;
 }
 
+bool operator==(const std::shared_ptr<Schema>& lhs, const std::shared_ptr<Schema> rhs) {
```

Comment:
I think we should avoid overriding operators for stl containers, especially ones which already have those operators. To do otherwise seems like a recipe for [unfortunate surprises](https://google.github.io/styleguide/cppguide.html#Operator_Overloading). That said, an overloaded equality operator which operates on `const Schema&` seems fine (though it can't do null checking)

Reply:
This is for testing, very handy to write `EXPECT_EQ(schema_1, schema_2)`. I can move this into test-utils.h, but @pitrou asked me to add this into the public interface in a different PR for a different class.

I think we should take a position on this.

Reply:
If it's for public usage, my comment holds. If it's for testing only, I'd prefer that we add `EXPECT_PTR_EQ` or so

Reply:
I agree with @bkietz that it looks like a bad idea.